### ISSUE TITLE
Orbot proxy gets frequently disabled

### ISSUE THREAD

**Issue by tzugen** [celzero/rethink-app#1065]
Hi, thanks for the work with the split proxies,
its mostly working good.

However, everytime there is some kind of connection issue and rethink looses the connection to orbot,
the whole proxy gets deactivated and there is no retry. Meaning that just a little fluke in the mobile network can cause the proxy setup to get disabled and everything being routed not trough tor.

The much safer option would be to keep the setting on, and if there is no connection to orbot possible, then the packets simply cannot be routed. People use Tor mostly for reasons of anonymity, so breaking is much prefered over keeping things running but leaking traffic over the wrong circuits.

**Comment by tzugen** [comment]
I will try to get some logs and pinpoint what triggers it. I am seeing it frequently enough..

**Comment by tzugen** [comment]
I have located the bug:
stopOrbot() is called also on failure, and this function, as a side-effects changes the proxy settings. 

```

    /**
     * Updates the shared preference related to Orbot Stop. Notifies the user about the Orbot
     * failure/stop. The notification will be sent if the user not initiated the stop.
     */
    fun stopOrbot(isInteractive: Boolean) {
        if (!isInteractive && selectedProxyType != AppConfig.ProxyType.NONE.name) {
            val notificationManager =
                context.getSystemService(VpnService.NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.notify(
                NOTIF_CHANNEL_ID_PROXY_ALERTS,
                ORBOT_SERVICE_ID,
                createNotification().build()
            )
        }
        selectedProxyType = AppConfig.ProxyType.NONE.name
        appConfig.removeAllProxies()
        persistentState.orbotConnectionStatus.postValue(false)
        context.sendBroadcast(getOrbotStopIntent())
        if (DEBUG) Log.d(LOG_TAG_VPN, "stop orbot, remove from proxy")
    }
```

https://github.com/celzero/rethink-app/blob/9ff4ad162f4246749b702c6e3892ce70a16cf90d/app/src/main/java/com/celzero/bravedns/util/OrbotHelper.kt#L257C37-L257C37

**Comment by tzugen** [comment]
To reproduce:
1. Configure Rethink to work with Orbot and confirm everything is working
1. Open Rethink Orbot Menu
2. Deactivate Orbot
3. Select "Stop Orbot" in the prompt
4. Orbot opens, stop it
5. Immidiately return to Rethink
6. Select "pure TCP Proxy" in the selection
7. The selection will immidiately jump back to None (Default) and a notification is shown about the failure.

**Comment by ignoramous** [comment]
> To reproduce:

This is on purpose since if Orbot is disconnected, the proxy is basically a blackhole. Previously, users complained about not stopping Rethink's proxy forwarder when Orbot was stopped...

**Comment by ignoramous** [comment]
I forgot: We do plan to introduce "lockdown" mode for proxies (that is, if a proxy is down, apps part of its tunnel lose connectivity and don't fallback on the underlying wifi/mobile network) as an opt-in: #976 

@tzugen Would that address your usecase?

**Comment by tzugen** [comment]
Yes kind of. But I believe this mode should be default for Tor instead of opt-in. I fail to see the case when you would want to oportunistally route some traffic through tor. You will end up with significant lower bandwidth with no real advantage for isolating different identities.

**Comment by tzugen** [comment]
btw, in general, I find it pretty annoying and against user expectations if apps modify settings themselves without asking.

**Comment by ignoramous** [comment]
> apps modify settings themselves without asking.

I understand. You could use the raw SOCKS5 proxy? It shouldn't turn OFF on its own. If it does, we'll fix that behaviour.

Note that, Orbot integration is exactly that. An integration. If Orbot is down, Rethink disconnects because it can't really "see" what's going on and has no choice but to quit.

**Comment by tzugen** [comment]
Thanks!

**Comment by tzugen** [comment]
Actually the workaround with using the proxy settings is not a real work around, because it seems i can only configure per-app proxy settings using the orbot ui

**Comment by ignoramous** [comment]
> seems i can only configure per-app proxy settings using the orbot ui

We are bringing this support in `v055b`, I believe. If not (as in we don't fix all the bugs, of which there are a few nagging ones), it may land in `v055c`.

**Comment by tzugen** [comment]
Hi, how is your approach to PRs and other developers. I would be up to support rethink more. I am maintaining two FLOSS android apps and think I could contribute good. But for me its important to have good communication around what is wanted and needs to be done, because its frustrating to put a lot of work into a changeset, but then its never merged.

**Comment by ignoramous** [comment]
>  I would be up to support rethink more.

Thanks (:

> Hi, how is your approach to PRs and other developers.

If it is anything drastic:

- Create an issue on GitHub.
- Wait for our approval.

If it is a minor / targeted bug fix:

- Send a pull request.
- Wait for us to comment or merge it in.

If it is a minor / important refactor:

- Send a pull request.
- Know that we ourselves refactor the code release from release. A lot.
- May or may not get merged.

> because its frustrating to put a lot of work into a changeset, but then its never merged.

Yeah, don't put work into a *large* changeset. I wouldn't recommend it unless you're deeply familiar with the codebase. Though, we've merged quite a few such changes by @HrBDev, @Uldiniad, and @Ch4t4r already, but they're already expert Android programmers in their own right.

**Comment by ignoramous** [comment]
> both my rethink and orbot are restarting on boot, however rethink disables orbot proxy since it seems to load later.
> 
> instead, i would ask the proxy in rethink be kept active, but just be used only when orbot is actually active too.

https://github.com/celzero/rethink-app/issues/1248#issue-2173398952
